src/promptflow-tools/tests/test_open_source_llm.py

```diff
@@ -0,0 +1,159 @@
+    def test_open_source_llm_con_model_chat(self, gpt2_custom_connection):
+        del gpt2_custom_connection.configs['model_family']
+        try:
+            os = OpenSourceLLM(gpt2_custom_connection)
+            os.call(self.completion_prompt, API.COMPLETION)
+            assert False
+        except OpenSourceLLMKeyValidationError:
+            pass
+
+    def test_open_source_llm_escape_chat(self):
+        danger = r"The quick \brown fox\tjumped\\over \the \\boy\r\n"
+        out_of_danger = ContentFormatterBase.escape_special_characters(danger)
+        assert out_of_danger == "The quick \\brown fox\\tjumped\\\\over \\the \\\\boy\\r\\n"
+
+    def test_open_source_llm_llama_parse_chat_with_chat(self):
+        LlamaContentFormatter.parse_chat(self.chat_prompt)
+
+    def test_open_source_llm_llama_parse_multi_turn(self):
+        multi_turn_chat = """user:
+You are a AI which helps Customers answer questions.
+
+What is the best movie of all time?
+
+assistant:
+Mobius, which starred Jared Leto
+
+user:
+Why was that the greatest movie of all time?
+"""
+        LlamaContentFormatter.parse_chat(multi_turn_chat)
+
+    def test_open_source_llm_llama_parse_system_not_accepted(self):
+        bad_chat_prompt = """system:
+You are a AI which helps Customers answer questions.
+
+user:
+""" + self.completion_prompt
+        try:
+            LlamaContentFormatter.parse_chat(bad_chat_prompt)
+            assert False
+        except OpenSourceLLMUserError:
+            pass
+
+    def test_open_source_llm_llama_parse_ignore_whitespace(self):
+        bad_chat_prompt = f"""system:
+You are a AI which helps Customers answer questions.
+
+user:
+
+user:
+{self.completion_prompt}"""
+        try:
+            LlamaContentFormatter.parse_chat(bad_chat_prompt)
+            assert False
+        except OpenSourceLLMUserError:
+            pass
+
+    def test_open_source_llm_llama_parse_chat_with_comp(self):
+        try:
+            LlamaContentFormatter.parse_chat(self.completion_prompt)
+            assert False
+        except OpenSourceLLMUserError:
+            pass
+
+    @pytest.mark.skip_if_no_key("gpt2_custom_connection")
+    def test_open_source_llm_llama_endpoint_miss(self, gpt2_custom_connection):
+        gpt2_custom_connection.configs['endpoint_url'] += 'completely/real/endpoint'
+        os = OpenSourceLLM(gpt2_custom_connection)
+        try:
+            os.call(
+                self.completion_prompt,
+                API.COMPLETION)
+        except OpenSourceLLMOnlineEndpointError:
+            pass
+
+    @pytest.mark.skip_if_no_key("gpt2_custom_connection")
+    def test_open_source_llm_llama_deployment_miss(self, gpt2_custom_connection):
+        os = OpenSourceLLM(gpt2_custom_connection)
+        try:
+            os.call(
+                self.completion_prompt,
+                API.COMPLETION,
+                deployment_name="completely/real/deployment-007")
+        except OpenSourceLLMOnlineEndpointError:
```

Comment:
A more common way to assert exception is `pytest.raises(<Exception>)`. You could refer to https://github.com/microsoft/promptflow/blob/6f767704eca7ac8f8bdea6a9b75931a463753c81/src/promptflow-tools/tests/test_common.py#L43C1-L46C68

Reply:
Great learning experience for me - thank you chjinche!!!